baseapp/baseapp.go

```diff
@@ -876,20 +881,22 @@ func (app *BaseApp) prepareProposal(req abci.RequestPrepareProposal) ([][]byte,
 	return txsBytes, nil
 }
 
-func (app *BaseApp) processProposal(req abci.RequestProcessProposal) error {
-	for _, txBytes := range req.Txs {
-		tx, err := app.txDecoder(txBytes)
-		if err != nil {
-			return err
-		}
+func (app *BaseApp) DefaultProcessProposal() sdk.ProcessProposalHandler {
+	return func(ctx sdk.Context, req abci.RequestProcessProposal) abci.ResponseProcessProposal {
+		for _, txBytes := range req.Txs {
+			tx, err := app.txDecoder(txBytes)
+			if err != nil {
+				panic(err)
+			}
 
-		_, _, _, _, err = app.runTx(runTxProcessProposal, txBytes)
-		if err != nil {
-			err = app.mempool.Remove(tx)
-			if err != nil && err != mempool.ErrTxNotFound {
-				return err
+			_, _, _, _, err = app.runTx(runTxProcessProposal, txBytes)
+			if err != nil {
+				err = app.mempool.Remove(tx)
```

Comment:
Weren't you removing the TX at `DeliverTx` time in a previous version of this patch?
What's the reason for removing the TX at `ProcessProposal` time? What if this is not the decided proposal?

Reply:
It removes invalid TX only, so if this tx does not pass the checks we remove it from the mempool

Reply:
Let's discuss this during tomorrow's sync up

Reply:
Ahh right, so when a honest process receives a proposal and executes it, any txs that are deemed invalid should still be included in the proposal. I think this is what @sergio-mena is alluding to. In other words, we should not reject or remove the tx from the mempool.

Reply:
The question is "What should Process Proposal do when `runTx` fails (returns an error)?"  Possible options:

1) Return REJECT without removing the failing tx.
2) Return REJECT after removing the failing tx.
3) Return ACCEPT without removing the failing tx.
4) Return ACCEPT after removing the failing tx.

Here we selected option (4).  You're suggesting option (3) @sergio-mena?  If we did this, then Process Proposal becomes a no-op, what would be the purpose of executing `runTx` if it has no effects on state or response?

We discussed this on a call 2 weeks ago and had said we think this case is pretty rare since PrepareProposal also checks for success from `runTx` and removes the tx from the proposal if it fails.  Given this we chose for a soft failure, just remove the tx and return ACCEPT.  The ABCI 1.0 docs for ProcessProposal also [warns against returning REJECT](https://github.com/tendermint/tendermint/blob/master/spec/abci%2B%2B/abci++_methods.md#processproposal).

> Moreover, application implementors SHOULD always set ResponseProcessProposal.status to ACCEPT, unless they really know what the potential liveness implications of returning REJECT are.

Reply:
Yeah, I realize I was vague in my comment here, and that led to misinterpretation. Sorry about that.
My comment was simply, "why bother"? The TX will eventually be removed by either `reCheckTx`, or a future `PrepareProposal` (if it's still invalid by then)